### PR TITLE
openssl dgst: add option to specify output length for XOF

### PR DESCRIPTION
**Pull request by ueno**
This adds the `-xoflen` option to control the output length of the XOF algorithms, such as SHAKE128 and SHAKE256.

I initially thought that this can be implemented as a more generic option (say `-mdopt xoflen:100`) that allows to specify the "xoflen" param with `EVP_MD_CTX_set_params`. However, in the end I realized that it nevertheless requires a special case for XOF for the following reasons:
- `dgst` makes use of `BIO` and the underlying `gets` method truncates the output
- in the default provider, the value of "xoflen" is looked up with `OSSL_PARAM_get_size_t` which only works on a parameter values stored with `OSSL_PARAM_construct_size_t`, not the one returned from `app_params_new_from_opts`

I'd like to hear any suggestions.

Fixes: #11964

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
Daiki, could you please submit a CLA? Otherwise the PR cannot be merged.

**Comment by t8m** [comment]
https://www.openssl.org/policies/openssl_icla.pdf

File: apps/dgst.c
Comment (t8m) [inline apps/dgst.c:407]: Please also check that sigkey is not specified for xoflen > 0.

File: apps/dgst.c
Comment (t8m) [inline apps/dgst.c:436]: This should not be necessary IMO. The following BIO_reset call should handle it fine.

**Review by t8m — CHANGES_REQUESTED** [review]

File: test/recipes/20-test_dgst.t
Comment (levitte) [inline test/recipes/20-test_dgst.t:128]: `\([^\)]*data.txt\)` can be changed to `\(\Q$testdata\E\)`, FYI

**Comment by paulidale** [comment]
>     * in the default provider, the value of "xoflen" is looked up with `OSSL_PARAM_get_size_t` which only works on a parameter values stored with `OSSL_PARAM_construct_size_t`, not the one returned from `app_params_new_from_opts`

This isn't quite correct.  The OSSL_PARAM helpers make an effort to convert between built in numeric types.  Thus, using other numeric construct functions (e.g. `OSSL_PARAM_construct_int`) will work here.  I.e. the parameter returned from `app_params_new_from_opts` should work and if it doesn't we've got a bug.

**Comment by ueno** [comment]
retriggering the CLA check

**Comment by t8m** [comment]
@ueno could you please address the comments from Richard and me?

**Comment by ueno** [comment]
I'm sorry for the delay; addressed those issues.

File: apps/dgst.c
Comment (ueno) [reply apps/dgst.c:442]: I agree; this is because the use of `EVP_DigestFinalXOF` in `do_fp` breaks the invariant where `md` is always set in the underlying ctx, when the BIO is marked as initialized, with `init = 1`.

**Comment by ueno** [comment]
> > ```
> > * in the default provider, the value of "xoflen" is looked up with `OSSL_PARAM_get_size_t` which only works on a parameter values stored with `OSSL_PARAM_construct_size_t`, not the one returned from `app_params_new_from_opts`
> > ```
> 
> This isn't quite correct. The OSSL_PARAM helpers make an effort to convert between built in numeric types. Thus, using other numeric construct functions (e.g. `OSSL_PARAM_construct_int`) will work here. I.e. the parameter returned from `app_params_new_from_opts` should work and if it doesn't we've got a bug.

I think this is because "xoflen" is defined as a bignum:
https://github.com/openssl/openssl/blob/master/providers/implementations/digests/sha3_prov.c#L261
while it is always accessed as size_t. Would it make sense to change the definition to use `OSSL_PARAM_size_t`?

File: apps/dgst.c
Comment (t8m) [reply apps/dgst.c:442]: Ha, this is actually a bug/API regression in 3.0.

Comment (t8m) [reply apps/dgst.c:442]: See #13402
So please remove that call and we should rebase this upon the #13402 once it is merged.

**Comment by paulidale** [comment]
> while it is always accessed as size_t. Would it make sense to change the definition to use `OSSL_PARAM_size_t`?

OSSL_PARAM_size_t uses OSSL_PARAM_UNSIGNED_INTEGER underneath.  OSSL_PARAM_UNSIGNED_INTEGER doesn't imply bignum -- they can always convert to bignum but could be native types too.

As I mentioned, the conversion between numeric types happens automatically.  Setting this via a double, int, unsigned int or size_t will all work (assuming the length is within range of these types).

**Comment by ueno** [comment]
If we were able to use OSSL_PARAM for this, I guess the patch would look something like:
https://github.com/ueno/openssl/commit/29abe6407538ab318468120dd2836d652afeca82

I'm new to OpenSSL, so let me know which approach is more palatable.

File: apps/dgst.c
Comment (ueno) [reply apps/dgst.c]: @t8m rebased against the master now that #13402 has been merged, though I still like to hear any opinions on https://github.com/openssl/openssl/pull/13245#issuecomment-729080953.

**Comment by t8m** [comment]
The test failure seems relevant.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
ping for second review

**Review by mattcaswell — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Merged to master. Thank you for the contribution.

**Pull request closed without merging**
